master/pkg/schemas/merge.go

```diff
@@ -0,0 +1,214 @@
+	if vObj.Kind() != reflect.Ptr {
+		panic("non-pointer in merge")
+	}
+	// obj can't be a nil pointer, because Merge(nil, y) doesn't make any sense.
+	if vObj.IsZero() {
+		panic("nil pointer in merge")
+	}
+
+	// *obj must have the same type as src.
+	assertTypeMatch(vObj.Elem(), vSrc)
+
+	// Think: *obj = merge(*obj, src)
+	vObj.Elem().Set(merge(vObj.Elem(), vSrc, name))
+}
+
+func assertTypeMatch(obj reflect.Value, src reflect.Value) {
+	if obj.Type() == src.Type() {
+		return
+	}
+	panic(
+		fmt.Sprintf(
+			"type mismatch in merge; can't merge %T into %T",
+			src.Interface(),
+			obj.Interface(),
+		),
+	)
+}
+
+// merge is the recursive layer under Merge.  obj and src must always have the same type, and the
+// return type will also be the same.  The return value will never share memory with src, so it is
+// safe to alter obj without affecting src after the fact.
+func merge(obj reflect.Value, src reflect.Value, name string) reflect.Value {
+	// fmt.Printf("merge(%T, %T, %v)\n", obj.Interface(), src.Interface(), name)
+	assertTypeMatch(obj, src)
+
+	// If src is nil, return obj unmodified.
+	if src.Kind() == reflect.Ptr || src.Kind() == reflect.Interface {
+		if src.IsZero() {
+			return cpy(obj)
+		}
+	}
+
+	// Handle nil pointers by simply copying the src.
+	if obj.Kind() == reflect.Ptr && obj.IsZero() {
+		return cpy(src)
+	}
+
+	// If the object is Mergable, we only call Merge on it and return it as-is.
+	if mergeable, ok := obj.Addr().Interface().(Mergable); ok {
+		mergeable.Merge(src.Interface())
+		return obj
+	}
+
+	switch obj.Kind() {
+	case reflect.Ptr:
+		// We already checked for nil pointers, so just recurse on the Elem of the value.
+		obj.Elem().Set(merge(obj.Elem(), src.Elem(), name))
+
+	case reflect.Struct:
+		// Detect what to do with union fields.  There are 4 important cases:
+		//  1. src has a union member, obj does not -> recurse into that field.
+		//  2. src has a union member, obj has the same one -> recurse into that field.
+		//  3. src has a union member, obj has the different one -> do not recurse.
+		//  4. src has no union member -> recursing is a noop and doesn't matter
+		// Logically, this reduces to:
+		//   - if obj has a union member, src does not have the same one -> don't recurse.
+		//   - else -> recurse
+		recurseIntoUnion := true
+		for i := 0; i < src.NumField(); i++ {
+			structField := src.Type().Field(i)
+			if _, ok := structField.Tag.Lookup("union"); ok {
+				if !obj.Field(i).IsZero() && src.Field(i).IsZero() {
+					recurseIntoUnion = false
+					break
+				}
+			}
+		}
+		// Recurse into each field of the struct.
+		for i := 0; i < src.NumField(); i++ {
+			structField := src.Type().Field(i)
+			if _, ok := structField.Tag.Lookup("union"); ok && !recurseIntoUnion {
+				continue
+			}
+			fieldName := fmt.Sprintf("%v.%v", name, structField.Name)
+			x := merge(obj.Field(i), src.Field(i), fieldName)
+			obj.Field(i).Set(x)
```

Comment:
This is what I was talking about in `fillDefaults` that I feel is missing there.